README.md
airtest
=====
this is python lib for airtest

AirTest PoPo Discuss Group: **1275211**

成功也好，失败也好，或者只想吐槽，欢迎直接反馈到群里。

关于**屏幕旋转**一直也没有一个很好的解决方案，现在的方法需要手工去设置旋转方向。
具体需要参考这篇文章 [airtest中的屏幕旋转如何设置](http://doc.mt.nie.netease.com/doku.php?id=airtest-screen-rotate)

airtest目前的wiki在, 很多问题可以再上面找到答案： <http://doc.mt.nie.netease.com/doku.php?id=airtest>

airtest具体的API依然维护这个README上面

[![Build Status](http://jenkins.mt.nie.netease.com/job/airtest_selftest/badge/icon)](http://jenkins.mt.nie.netease.com/job/airtest_selftest/)

## INSTALL
[Click HERE](INSTALL.md)

## Check if install OK
run `air.test` in terminal.

## DEMOS
Run [flappybird example program](http://git.mt.nie.netease.com/hzsunshx/flappybird)

download through: `git clone http://git.mt.nie.netease.com/hzsunshx/flappybird`

## LIMIT
[airtest的限制](LIMIT.md)

## START USING AIRTEST
### step1: generate air.json file.
run this command to generate file `air.json`
```
air.test gen $apk_file_path
```

### step2: prepare test code
main.py code example
```
import airtest

devno = os.getenv('AIRTEST_DEVNO') or 'xxxxxx'  # get from: adb devices
appname = os.getenv('AIRTEST_APPNAME') or 'com.netease.xxxx'
device =  airtest.ANDROID

app = airtest.connect(devno, appname=appname, device=device)
print app.shape() # show device size
```

take screen snapshot by run: `air.test snapshot`, screen will save to `screen.png`
cut the image part from it.

还可以使用airtest专用的gui，来完成快速的截图和代码调试，[点击安装地址](http://git.mt.nie.netease.com/hzsunshx/airtest-gui)

### step3: run test code
1. connect you device. if you use android, `adb devices` should see a devices
2. install app. `air.test install`
3. runtest. `air.test runtest`
4. generate log, and show html in Chrome. `air.test log2html --listen --port=8888 report`
5. uninstall app. `air.test uninstall`

open browser, input `127.0.0.1:8888`. You should see html report now.

## API Reference
after installed successfully. you can import like
```
# import python lib
import airtest
```

step1 connect device
```
# get serialno by call: adb devices
devno = os.getenv('AIRTEST_DEVNO') or 'xxxxx888882111' # <win_file_name | mac bundleid | android serialno>
appname = os.getenv('AIRTEST_APPNAME') or 'com.netease.rz' # app name
deviceType = 'android' # can be windows or ios

# connect to your android devices
# default value: device='android', monitor=True
app = airtest.connect(devno, appname=appname, device=deviceType, monitor=True)
```

```
app.monitor.start() # start monitor
app.monitor.stop() # stop monitor
app.globalSet(monitor_interval=1.0) # set monitor interval to 1s
```

takeSnapshot(filename) # filename show with extention (.jpg or .png)
```
app.takeSnapshot('snapshot.png')
```

click(P, [seconds], eventType=airtest.EV_DOWN_AND_UP) # click by image file
```
app.click(P)
# P can be
# - filename: 'start.png'
# - position: (100, 200)
# - percent: (0.1, 0.02)    # equal to (width*0.1, height*0.02)

# click-timeout(only avaliable when P is string)
# equals to app.click(app.find('start.png', 20.0))
app.click('start.png', 20.0) # if start.png not found in 20s, Exception will raised.

# long press (this is an experimental function, report problem if something goes wrong)
app.click(P, eventType=airtest.EV_DOWN)
time.sleep(1.0) # press 1s
app.click(P, eventType=airtest.EV_UP)
```

find(...) # find a image position located in screen
```
(x, y) = app.find(filename)
```

findall(self, imgfile, maxcnt=None, sort=None): # sort = <None|"x"|"y">
```
findall('start.png', maxcnt=2)
findall('start.png', maxcnt=2, sort='x') # sort ordered by x row
```

wait(...) # wait until image shows
```
app.wait(filename, [seconds])
# filename is the image name
# seconds is the longest time waited.
# @return position images find, or Raise RuntimeError
# this is called find(..) to get images position
```

safeWait(filename, [seconds]) # like wait, but don't raise RuntimeError
```
app.safeWait(filename) # return None if not found, else point
```

exists(...) # judge if image exists
```
app.exists('apple.png')
# @return (True|False)
# just exactly call wait
```

drag(...) # drag one place to and onother place
```
app.drag(fromP, toP)
# fromP, toP: like click param, can be filename or position
```

shape() # get screen size(width and height)
```
(w, h) = app.shape()
# return width and height
```

```
# example of drag from left to right
(x1, y1), (x2, y2) = (w*0.2, h*0.5), (w*0.8, h*0.5)
app.drag((x1,y1), (x2,y2))
```

type(...) # type text
```
app.type('www.baidu.com\n') # type text and call keyevnet ENTER
```

keyevent(not recommemd to use now)
```
# back and menu(only for android)
app.keyevent('BACK')
app.keyevent('MENU')
```

### airtest settting
Mobile phone has 4 directions: `UP,DOWN,LEFT,RIGHT`.
Change rotation through. more info view: <http://doc.mt.nie.netease.com/doku.php?id=airtest-screen-rotate>
```
app.globalSet({'rotation': 'RIGHT'})
```

Change image recognize sensitivity
```
# threshold range [0, 1)
# the bigger the accurate. If set to 1, then app can't recognize anything
app.globalSet({'threshold': 0.3}) 
```

关于更详细的配置，请参考: <http://doc.mt.nie.netease.com/doku.php?id=airtest#设置接口globalset和globalget>

## 更新日志
[CHANGELOG](CHANGELOG.md)

## Team

	孙圣翔 hzsunshx@corp.netease.com
	- 项目管理员
	
	刘欣 gzliuxin@corp.netease.com
	- IOS后端接口开发者

	熊博 gzxiongbo@corp.netease.com
	- IOS后端接口开发者

	贾强槐 hzjiaqianghuai@corp.netease.com
	- 图像识别

## development document
AndroidKeyMap: <http://developer.android.com/reference/android/view/KeyEvent.html>

used by call, eg
```
app.keyevent("HOME")
```


## License
This project is under the MIT License. See the [LICENSE](LICENSE) file for the full license text.


CHANGELOG.md
Change Log
=================

### 2014/07/17
1. 增加start,stop,clear函数
1. globalSet增加mem和cpu监控周期调整。globalSet({'monitor_interval': 5})

### 2014/07/21
1. 增加airtest自测代码
2. 修正ios中的屏幕旋转问题
3. 增加关于屏幕设置的文档

### 2014/07/31
1. dev.getMem接口调整，原来返回PSS值，现在返回{'PSS':pss, 'VSS':vss, 'RSS':rss}
2. 新增airtest的图形界面，方面快速的截图和写代码
3. 图像文件名支持unicode中文，支持在多个图片文件夹里自动搜索
4. 增加globalGet函数

### 2014/08/04
1. globalSet更新,支持globalSet(key=value)，兼容原有的globalSet({})

### 2014/08/12
1. 增加image_match_method配置项，可以手动选择图形匹配算法
2. 环境变量AIRTEST_DEVNO改名为AIRTEST_PHONENO(影响范围，使用air.test命令行的runtest的程序)

### 2014/08/14
1. 增加app.monitor变量， 可以通过app.monitor.start() 启动监控和通过app.monitor.stop()停止监控
2. 增加fps的支持

### 2014/08/20
1. 新增airtoolbox工具（该工具可以运行在手机上，采集手机的内存，网卡信息，模拟长按的操作）
2. 完成air.test gen命令

### 2014/08/27
1. 支持python -mairtest 等价于 命令行的air.test
2. log2html生成的文件支持点击放大的效果。
3. 修复air.test gen命令，对于某些apk，不生效的问题。

### 2014/08/29
1. 输入法支持中文及特殊字符

### 2014/09/16
1. 支持输入法的自动安装
2. 图像识别算法，分离出feature SIFT算法
3. 修正屏幕点亮的问题。多余代码清理
4. 易用性修改，如果只有一个android设备连接，可以不用指定serialno


airtest/cli.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

#!/usr/bin/env python
# coding: utf-8
#

'''
phone(android|iphone) autotest framework. gen can generate air.json
Usage:
    air.test gen <apk>
    air.test (runtest|install|uninstall) [-p PLATFORM] [SERIALNO]
    air.test log2html [--listen] [--port=PORT] <HTMLDIR>
    air.test snapshot [-p PLATFORM] [-r ROTATION] [SERIALNO]
    air.test all [--steps STEPS] [-H HTMLDIR] [-p PLATFORM] [SERIALNO]
    air.test update

Options:
    -h --help       Show this screen
    -p PLATFORM     android or iphone [default: android]
    -s SERIALNO     Specify devices serialno(needed)
    --steps STEPS   the steps one by one [default: install,runtest,log2html,uninstall]
    -H HTMLDIR      Save html report
    --port PORT     for log2html open a webserver to view report [default: 8888]
    -r ROTATION     rotation of device, one of UP,DOWN,LEFT,RIGHT
'''

__version__ = '0.1.0702'

import json
import sys
import os
import urllib
import subprocess
import re

from docopt import docopt
# from com.dtmilano.android.viewclient import ViewClient 

import airtest
from airtest.base import exec_cmd
from airtest import log2html, jsonlog

def urlretrieve(url, filename=None):
    print 'DOWNLOAD:', url, '->', filename
    return urllib.urlretrieve(url, filename)

F = {} #json.load(open(jsonfile, 'r'))
SUBPROCESS = []
platform = 'android'
serialno = None
rotation = 'UP'

def xpath(*paths):
    v=F
    for p in paths:
        v = v.get(p, {})
    return v if v else None

def run_snapshot():
    global serialno
    if not serialno and platform=='android':
        print F
        ds = airtest.getDevices()
        print ds
        if len(ds) == 1:
            serialno=ds[0][0]
        else:
            sys.exit("too many devices, don't know which you want")
    assert serialno != None
    print 'snapshot', serialno
    app = airtest.connect(serialno, device=platform)
    app.globalSet(dict(tmpdir='.'))
    print 'ROTATION:', rotation
    app.globalSet(dict(rotation=rotation))
    app.takeSnapshot('screen.png')
    print 'screenshot save to "screen.png"'
    #if platform == 'android':
    #    c, _ = ViewClient.connectToDeviceOrExit(serialno=serialno)
    #    c.takeSnapshot().save('screen.png')
    #    print 'screenshot save to "screen.png"'
    #else:
    #    print 'not supported:', platform

def run_gen():
    airjson ={
      "cmd": "python main.py",
      "android": {
        "apk_url": "http://10.246.13.110:10001/demo-release-signed.apk",
        "package": "com.netease.xxx",
        "activity": "main.activity"
      }
    }
    with open('air.json', 'w') as file:
        json.dump(airjson, file, indent=4)
    print 'nice'

def run_install():
    if platform == 'android':
        apk_url = xpath(platform, 'apk_url')
        if re.match('^\w{1,2}tp://', apk_url):
            urlretrieve(apk_url, 'test.apk')
            apk_url = 'test.apk'
        exec_cmd('adb', '-s', serialno, 'install', '-r', apk_url)
        package, activity = xpath(platform, 'package'), xpath(platform, 'activity')
        exec_cmd('adb', '-s', serialno, 'shell', 'am', 'start', '-n', '/'.join([package, activity]), timeout=10)
    else:
        print 'not supported:', platform

def run_uninstall():
    if platform == 'android':
        exec_cmd('adb', '-s', serialno, 'uninstall', xpath(platform, 'package'))
    else:
        print 'not supported:', platform

def run_runtest():
    env = {
            'SERIALNO': serialno, 
            'AIRTEST_PHONENO': serialno, # not suggested, delete me in 2014/10/01
            'AIRTEST_DEVNO': serialno,
            'AIRTEST_APPNAME': xpath(platform, 'package')}
    exit_code = exec_cmd(xpath('cmd'), timeout=30*60, shell=True, env=env)
    assert exit_code == 0

def run_log2html():
    if F.get('logfile') and F.get('htmldir'):
        log2html.render(F.get('logfile'), F.get('htmldir'))
        if F.get('listen'):
            p = subprocess.Popen(['python', '-mSimpleHTTPServer', F.get('port')], stdout=sys.stdout, stderr=sys.stderr, cwd=F.get('htmldir'))
            SUBPROCESS.append(p)
            p.wait()
            #os.system('cd %s; python -mSimpleHTTPServer %s' %(F.get('htmldir'), F.get('port')))

def run_update():
    print 'run this command manualy'
    print 'pip', 'install', '--upgrade', 'git+http://git.mt.nie.netease.com/hzsunshx/airtest.git'

def main():
    global F, platform, serialno, rotation
    arguments = docopt(__doc__, version='0.1')

    # set action
    action=''
    for act in ['all', 'install', 'log2html', 'runtest', 'snapshot', 'uninstall', 'update', 'gen']:
        if arguments.get(act):
            action = act
            break
    if not action:
        sys.exit('No action specified, see --help')
    print 'RUN action:', action

    # handle command in console
    cmd_name = sys.argv[1]
    from airtest import console
    func = console.COMMANDS.get(cmd_name)
    if func:
        return func.main(*sys.argv[2:])
    # else:sys.exit('cmd(%s) not exists' % cmd_name)

    # load conf
    cnf = 'air.json'
    if os.path.exists(cnf):
        F = json.load(open(cnf))
    if not 'logfile' in F:
        logfile = 'log/airtest.log'
        F['logfile'] = logfile
        os.environ['AIRTEST_LOGFILE'] = logfile

    F['htmldir'] = arguments.get('<HTMLDIR>') or arguments.get('-H')
    F['port'] = arguments.get('--port')
    F['listen'] = arguments.get('--listen')

    if action in ['log2html', 'update']:
        return globals().get('run_'+action)()

    # check phoneno and platform
    serialno = arguments['SERIALNO']
    platform = arguments.get('-p', 'android')
    rotation = arguments.get('-r', 'UP')
    if platform == 'android' and not arguments.get('SERIALNO'):
        devices = [dev for dev in airtest.getDevices() if dev[1] != 'unknown']
        if len(devices) != 1:
            sys.exit("can't determine which devices to use, please run: 'adb devices'")
        arguments['SERIALNO'] = devices[0][0]
        serialno = arguments['SERIALNO']

    if action in ['snapshot']:
        return globals().get('run_'+action)()

    print 'PREPARE platform: %s' %(platform)
    print 'PREPARE serialno: %s' %(serialno)
    #exec_cmd('adb', 'start-server', timeout=10)

    #print arguments
    if not os.path.exists(cnf) and action != 'gen':
        sys.exit('config file require: %s' %(cnf))
    if action == 'all':
        exitcode = 0
        for step in arguments['--steps'].split(','):
            fn = globals().get('run_'+step)
            if not fn or not callable(fn):
                sys.exit('no such step: %s' %(step))
            print 'STEP:', step
            try:
                fn()
            except Exception as e:
                jsonlog.JSONLog(logfile).writeline({'type':'cli', 'step':step, 'result':'failed', 'detail': str(e)})
                exitcode=1

        sys.exit(exitcode)
        return
    if action in ['install', 'uninstall', 'log2html', 'runtest']:
        print 'RUN:', action
        return globals().get('run_'+action)()


if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        for p in SUBPROCESS:
            p.kill()
        print 'Exited by user'
